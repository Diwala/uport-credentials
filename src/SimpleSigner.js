
export default function SimpleSigner(privateKey) {
  return function (data, callback) {
    // TODO implement signing
  }
}