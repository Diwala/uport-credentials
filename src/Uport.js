import { createJWT, verifyJWT } from './JWT'
import UportRegistry from 'uport-registry'

const INFURA_ROPSTEN = 'https://ropsten.infura.io'
const UPORT_REGISTRY_ADDRESS = '0xb9C1598e24650437a3055F7f66AC1820c419a679'

export default class Uport {
  constructor (settings) {
    this.settings = settings
    if (!this.settings.registry) {
      const registry = new UportRegistry()
      // The only function we're really interested from uport registry is getAttributes
      this.settings.registry = registry.getAttributes.bind(registry)
    }
  }

  // Create request token
  request (payload) {
    return createJWT(this.settings, {...payload, type: 'shareReq'})
  }

  // Receive response token from user and return data to promise
  receive (token) {
    return verifyJWT(this.settings, token).then(({payload, profile}) => (
      {...profile, ...(payload.own || {})}
    ))
  }

  // Create attestation
  attest ({sub, claim, exp}) {
    return createJWT(this.settings, {sub, claim, exp})
  }

  // send push 
  pushTo (pushToken, data, callback) {

  }
}
