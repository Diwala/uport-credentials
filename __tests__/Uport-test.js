import Uport from '../src/Uport'
import SimpleSigner from '../src/SimpleSigner'
import { SECP256K1Client, TokenVerifier, decodeToken } from 'jsontokens'

const privateKey = '278a5de700e29faae8e40e366ec5012b5ec63d36ec77e8a2417154cc1d25383f'
const publicKey = SECP256K1Client.privateKeyToPublicKey(privateKey)
const signer = SimpleSigner(privateKey)
const verifier = new TokenVerifier('ES256K', publicKey)
const uport = new Uport({signer, address: '0x001122'})

it('creates a valid JWT for a request', () => {
  return uport.request({requested: ['name', 'phone']}).then((jwt) => {
    return expect(verifier.verify(jwt)).toBeTruthy()
  })
})

it('has correct payload in JWT for a request', () => {
  return uport.request({requested: ['name', 'phone']}).then((jwt) => {
    return expect(decodeToken(jwt)).toMatchSnapshot()
  })
})
