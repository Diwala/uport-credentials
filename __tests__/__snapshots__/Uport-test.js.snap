exports[`test has correct payload in JWT for a plain request for public details 1`] = `
Object {
  "header": Object {
    "alg": "ES256K",
    "typ": "JWT",
  },
  "payload": Object {
    "iat": 1485321133996,
    "iss": "0x001122",
    "type": "shareReq",
  },
  "signature": "x5wOEbEa8gQtKtfteSDHkkIaW30kZmI9I8EQLk8-DmxL9GteEa9hvIAi8Zhxo22FPDFErxIpp48Ckgp9_ZsIEA",
}
`;

exports[`test has correct payload in JWT for a request 1`] = `
Object {
  "header": Object {
    "alg": "ES256K",
    "typ": "JWT",
  },
  "payload": Object {
    "iat": 1485321133996,
    "iss": "0x001122",
    "requested": Array [
      "name",
      "phone",
    ],
    "type": "shareReq",
  },
  "signature": "831teGtcqKacto5LDMLVURpCHxUN9Vvx2IN0-b05Q5gU3Md1wiuRDM91tP5Q0AyYTr1RTnkoYwmMcNH7wke7oA",
}
`;

exports[`test has correct payload in JWT for an attestation 1`] = `
Object {
  "header": Object {
    "alg": "ES256K",
    "typ": "JWT",
  },
  "payload": Object {
    "claim": Object {
      "email": "bingbangbung@email.com",
    },
    "exp": 1485321134996,
    "iat": 1485321133996,
    "iss": "0x001122",
    "sub": "0x112233",
  },
  "signature": "-mEzVMPYnzqFhOr0O7fs71-dWAacnllVyOdWQY0zh2ZdIt7-30IYTewds4tGlkLmMky-Y1ZjRmIsxmM7xvAgxg",
}
`;
